ripser/ripser.py

```diff
@@ -125,11 +225,29 @@ def ripser(
                 "The input point cloud has more columns than rows; "
                 + "did you mean to transpose?"
             )
-        X = pairwise_distances(X, metric=metric)
 
-    if not (X.shape[0] == X.shape[1]):
-        raise Exception("Distance matrix is not square")
-    dm = X
+    idx_perm = np.arange(X.shape[0])
```

Comment:
There are a lot of forks right here. Can we put all the exception handling at the top all in one place? Then have this be just logic for getting dm?

What about an interface like:

```
if n_perm and n_perm > X.shape[0]:
    raise Exception
if n_perm and distance_matrix and sparse.issparse(X):
   raise Exception

if n_perm:
    dm, r_cover, idx_perm = _compute_permuted_distance_matrix(X, N=n_perm, distance_matrix=distance_matrix, metric=metric)
else:
    dm = X if distance_matrix else pairwise_distances(X, metric=metric)
    r_cover, idx_perm = 0.0, np.arange(X.shape[0])
```